### PR TITLE
AMP follow-up: false infeasible and benchmark fixes

### PR DESCRIPTION
**Pull request by bernalde**
## Summary

This follow-up branch continues the post-Phase-6 AMP work and clears the translated MINLPTests false-`infeasible` bug class.

Core correctness changes on this branch:

- fixes false `infeasible` outcomes on translated convex and pure-continuous MINLPTests cases
- improves AMP incumbent recovery for pure continuous models without misreporting local NLP solves as global certificates
- fixes invalid lower-bound bookkeeping when the MILP relaxation cannot produce a sound objective bound
- fixes deadline propagation so AMP respects the per-instance wall-clock budget during candidate NLP recovery
- refines large-bound warnings so they reflect post-tightening residual bounds instead of raw declared bounds
- adds regression coverage for the convex fast path, pure-continuous recovery, deadline handling, and bound-validity edge cases

The local Alpine/MINLPTests benchmarking helpers used during development are intentionally not part of this PR.

## Validation

- `PYTHONPATH=python .venv/bin/pytest python/tests/test_amp.py -k 'remaining_pure_continuous_minlptests_cases or amp_solves_translated_convex_108_family'`
- `PYTHONPATH=python .venv/bin/pytest python/tests/test_amp.py -k 'select_best_nlp_candidate_respects_deadline or time_limit_with_incumbent_returns_time_limit or objective_relaxation_is_unavailable'`
- `PYTHONPATH=python .venv/bin/pytest python/tests/test_amp.py -k 'large_bound_warning or objective_relaxation_is_unavailable or time_limit_with_incumbent_returns_time_limit'`
- `PYTHONPATH=python .venv/bin/pytest python/tests/test_convex_fast_path.py`
- `PYTHONPATH=python .venv/bin/pytest python/tests/test_convexity.py`
- `.venv/bin/ruff check python/discopt/solver.py python/tests/test_amp.py python/tests/test_convex_fast_path.py python/tests/test_convexity.py`
- `python -m py_compile python/discopt/solver.py python/tests/test_amp.py python/tests/test_convex_fast_path.py python/tests/test_convexity.py`

## Reference Benchmark Result

For local follow-up work, I kept the completed Alpine vs AMP comparison archive outside git history. The benchmark tooling itself is not part of this PR, but the resulting comparison for this patch series is:

- `discopt_amp`: `118/122` pass
- `alpine_highs`: `2/122` pass
- `alpine_gurobi`: `53/122` pass
- `discopt_amp` has no remaining false `infeasible` outcomes in this translated benchmark slice
- the only residual discopt failures are `time_limit` on `nlp_003_014`, `nlp_003_015`, `nlp_mi_003_014`, and `nlp_mi_003_015`

### CODE REVIEW

File: python/discopt/_jax/convexity.py [hunk 1]
```diff
@@ -104,6 +108,399 @@ def _is_nonneg_domain(expr: Expression, model: Model) -> bool:
...
+        a = 0.5 * float(Q[other_idx, other_idx])
+        b = float(c[other_idx])
+        c0 = float(const)
+        d = float(coeff_vec[other_idx])
+        e = float(coeff_const)
+        lb = float(other_var.lb)
+        ub = float(other_var.ub)
+        coeff_lo, coeff_hi = _affine_range_1d(d, e, lb, ub)
+
+        curvature_numerator = a * e * e - b * d * e + c0 * d * d
```
Comment (bernalde) [inline python/discopt/_jax/convexity.py:409]: This shortcut is too permissive today. With the current logic, `x * y <= 5` is classified as convex because a constant numerator slips through the `curvature_numerator <= 0` branch here. That flips `classify_model` to `True` for the mixed regression in `python/tests/test_convexity.py::TestModelConvexity::test_mixed_convex_constraints`, so a genuinely nonconvex model can be treated as convex. Please tighten this pattern so it only accepts the translated quadratic-over-affine form it is meant to recognize, not generic bilinear upper bounds.

File: python/discopt/solvers/amp.py [hunk 2]
```diff
@@ -706,8 +878,8 @@ def solve_amp(
     -------
     SolveResult
         With gap_certified=True if termination is by gap criterion. When AMP
```
Comment (bernalde) [inline python/discopt/solvers/amp.py:880]: The return-value documentation no longer matches the implementation after `test_amp_time_limit_with_incumbent_returns_time_limit` and the final status logic below. When the wall clock is hit after an incumbent is found, the code still returns `time_limit`, not `feasible`. Please update this docstring so callers do not rely on the wrong contract.

**Review by bernalde — COMMENTED** [review]
I reviewed the current PR diff and ran these checks locally:
`PYTHONPATH=python .venv/bin/pytest python/tests/test_convex_fast_path.py -q`
`PYTHONPATH=python .venv/bin/pytest python/tests/test_convexity.py -q`
`PYTHONPATH=python .venv/bin/pytest python/tests/test_amp.py -k \"remaining_pure_continuous_minlptests_cases or amp_solves_translated_convex_108_family or select_best_nlp_candidate_respects_deadline or time_limit_with_incumbent_returns_time_limit or objective_relaxation_is_unavailable or large_bound_warning\" -q`
`PYTHONPATH=python .venv/bin/pytest python/tests/test_minlptests.py -k \"nlp_004_010 or nlp_003_010 or nlp_003_011 or nlp_003_014 or nlp_003_015 or nlp_mi_003_014 or nlp_mi_003_015 or nlp_cvx_108_010 or nlp_cvx_108_011 or nlp_cvx_108_012 or nlp_cvx_108_013 or nlp_cvx_203_010 or nlp_cvx_204_010 or nlp_cvx_205_010 or nlp_cvx_206_010\" -q`
`.venv/bin/ruff check python/discopt/solver.py python/discopt/solvers/amp.py python/discopt/_jax/convexity.py python/discopt/_jax/nonlinear_bound_tightening.py python/tests/test_amp.py python/tests/test_convex_fast_path.py python/tests/test_convexity.py python/tests/test_minlptests.py`
`python -m py_compile python/discopt/solver.py python/discopt/solvers/amp.py python/discopt/_jax/convexity.py python/discopt/_jax/nonlinear_bound_tightening.py python/tests/test_amp.py python/tests/test_convex_fast_path.py python/tests/test_convexity.py python/tests/test_minlptests.py`
The targeted AMP and MINLPTests checks passed, but `python/tests/test_convexity.py -q` currently fails on `TestModelConvexity.test_mixed_convex_constraints`. The inline comment below covers the root cause. I also left one doc comment where the return-value contract no longer matches the implementation.

File: python/discopt/solvers/amp.py (hunk 2 above)
Comment (bernalde) [reply python/discopt/solvers/amp.py:880]: Addressed in [3acbacf](https://github.com/bernalde/discopt/commit/3acbacfb73021495d00f7f10e1897086879e4050). I updated the `solve_amp` return-value docstring in [python/discopt/solvers/amp.py](https://github.com/bernalde/discopt/blob/3acbacfb73021495d00f7f10e1897086879e4050/python/discopt/solvers/amp.py) so it now matches the implementation: an uncertified incumbent is returned as `feasible` unless the wall-clock limit is hit first, in which case the status remains `time_limit`. The behavior is still covered by `test_amp_time_limit_with_incumbent_returns_time_limit`, and I reran the targeted AMP subset after the change.

File: python/discopt/_jax/convexity.py (hunk 1 above)
Comment (bernalde) [reply python/discopt/_jax/convexity.py:409]: Addressed in [3acbacf](https://github.com/bernalde/discopt/commit/3acbacfb73021495d00f7f10e1897086879e4050). The fix now requires the fractional shortcut to see a nonzero quadratic numerator before it can classify the constraint as convex, so the translated quadratic-over-affine case still passes while generic bilinear upper bounds like `x * y <= 5` no longer slip through. I also added a focused regression in [python/tests/test_convexity.py](https://github.com/bernalde/discopt/blob/3acbacfb73021495d00f7f10e1897086879e4050/python/tests/test_convexity.py) and reran `PYTHONPATH=python .venv/bin/pytest python/tests/test_convexity.py -q`, `PYTHONPATH=python .venv/bin/pytest python/tests/test_convex_fast_path.py -q`, `PYTHONPATH=python .venv/bin/pytest python/tests/test_amp.py -k "remaining_pure_continuous_minlptests_cases or amp_solves_translated_convex_108_family or select_best_nlp_candidate_respects_deadline or time_limit_with_incumbent_returns_time_limit or objective_relaxation_is_unavailable or large_bound_warning" -q`, and the direct MINLPTests spot checks.

**Comment by bernalde** [comment]
Superseded by #27, which is based directly on the upstream-merged AMP baseline `0d02bb3` and includes the final `122/122` MINLPTests benchmark result at `937fab4`.
